rfc/system/4358-tuple-idl.md

```diff
@@ -0,0 +1,284 @@
+    ```proto
+    message TupleFieldType {
+        string name = 1;
+        LiteralType type = 2;
+    }
+
+    message TupleType {
+        string tuple_name = 1;
+        repeated TupleFieldType fields = 2;
+    }
+
+    message LiteralType {
+        oneof type {
+            // ...
+            TupleType tuple_type = 11;
+        }
+        // ...
+    }
+    ```
+
+## 2 Motivation
+
+Flytekit currently restricts the support of tuple types, which are commonly used in Python. This RFC aims to add support for tuple types in FlyteIDL and Flytekit.
+
+Note: Several issues regarding `NamedTuple` and `tuple` support in Flytekit have been identified:
+
+- [#1337](https://github.com/flyteorg/flyte/issues/1337)
+- [#3158](https://github.com/flyteorg/flyte/issues/3158)
+- [#4358](https://github.com/flyteorg/flyte/issues/4358)
+
+This feature aims to resolve all these issues.
+
+## 3 Proposed Implementation
+
+### FlyteIDL
+
+#### Type of NamedTuple and Tuple
+
+- To support NamedTuple, we need to store the name of each field and the name of the tuple itself. We will create a new message type `TupleFieldType` to store the name and type of each field. For `Tuple`, we can use the same `TupleType`, but the `name` and `tuple_name` will be empty strings.
+
+    ```proto
+    message TupleFieldType {
+        string name = 1;
+        LiteralType type = 2;
+    }
+
+    message TupleType {
+        string tuple_name = 1;
+        repeated TupleFieldType fields = 2;
+    }
+
+    message LiteralType {
+        oneof type {
+            // ...
+            TupleType tuple_type = 11;
+        }
+        // ...
+    }
+    ```
+
+- We could also consider supporting `typing.Tuple[T1, ...]` in the future by adding an additional field `is_univariate` in `TupleType`. However, the performance of `typing.Tuple[T1, ...]` is closely related to `typing.List[T1]`, so this use case could be covered by `LiteralType_CollectionType` for now. (#TO_DISCUSS)
+
+    ```proto
+    message TupleType {
+        string tuple_name = 1;
+        bool is_univariate = 2;
+        repeated TupleFieldType fields = 3;
+    }
+    ```
+
+- The above code should be added to [flyteidl/protos/flyteidl/core/types.proto](https://github.com/flyteorg/flyte/blob/master/flyteidl/protos/flyteidl/core/types.proto).
+
+#### Literal of NamedTuple & Tuple
+
+- Since the type, including the names of the fields and the tuple, is stored in `LiteralType`, we only need to store the value of each field in the tuple. We can use the existing `LiteralCollection` message from [flyteidl/protos/flyteidl/core/types.proto](https://github.com/flyteorg/flyte/blob/master/flyteidl/protos/flyteidl/core/literals.proto) to store the value of each field.
+
+    ```proto
+    message LiteralCollection {
```

Comment:
I think we still need a new literal for tuple. Otherwise, we can't properly get the literal type from the literal [here](https://github.com/flyteorg/flyte/blob/1cd775a417e4280fd78c9334f7e9329e53b017d3/flytepropeller/pkg/compiler/validators/utils.go#L247).